### DISCUSSION TITLE
Add feature for resolving individual users in a group for assigneesSampleSize

### DISCUSSION THREAD

**Discussion by bbckr** [renovatebot/renovate#25300] [Suggest an Idea]
### Tell us more.

Hey there 👋

For a given CODEOWNERS file that associates a Group to a dependency file that Renovate is making changes to, it would be nice to have a third configuration setting for allowing Renovate to expand the groups into individual users, so that MR assignees can be sampled from the list of individual users maintained in the group, rather than just assigning the entire group.

This is particularly helpful in large repositories where multiple Groups own different sections of the codebase, and the CODEOWNERS file only references Groups rather than individual users.

Furthermore, it allows Renovate MRs to spread the load of getting the MR reviewed and merged amongst a smaller sample size of users in the group, rather than entire group (which is subject to the bystander effect).

This is something that would be extremely helpful for our current adoption of Renovate on our much larger, shared monorepos.

Thanks for taking the time to consider this, and looking forward to hearing back on what yall think 🙏

**Comment by rarkins** [comment]
Which platform? e.g. github, gitlab, etc? Some platforms like github do the assignment automatically so Renovate has no say in it

**Reply by colinodell** [reply]
The gitlab platform's `addReviewers()` method has the ability to expand any groups into the full list of members (within Renovate) via `getMemberUserIDs()`.

What if Renovate did something similar, but for assignees, and where that expansion could be configured to happen _before_ assignee sampling occurs?  In other words, given this current code: 

https://github.com/renovatebot/renovate/blob/58cc2c073738daca066ed87130099f59630c6d9f/lib/workers/repository/update/pr/participants.ts#L41-L51

The `prepareParticipants()` method called on line 48 could optionally (through a new config option) expand any groups in the `assignees` list so that the sampling on line 50 would always select individual users and never a full group.

Currently, sampling might choose a large group from `assignees`, and thus everyone in that group would get assigned.  The proposed feature here would instead spread the load throughout that group by selecting just one or two people from that group at random.

I have not looked at how feasible this might be for other platforms, though.

**Reply by bbckr** [reply]
@rarkins  For us, this would be for Gitlab (see Colin's post above)

**Reply by rarkins** [reply]
If someone can find a way to do this elegantly, we would accept the PR. I think it would make most sense as a new option called something like `expandCodeownersGroups` which defaults to `false`

**Reply by bbckr** [reply]
Awesome, we can try to work on this then and will link once ready.

**Reply by stdedos** [reply]
It seems to be a problem also with `reviewersFromCodeOwners`.

... Also, if it somehow affects implementation, we are thinking to use it so that `assigneesFromCodeOwners: true`, `reviewersFromCodeOwners: true` - so there might be some collision 😅 (rationale: one from the team is supposed to own the bot's MR, and someone is supposed to review it)

Is there any progress with this?

**Reply by stdedos** [reply]
... and a stacktrace, if it helps somehow

```
  ▽ err: {message: "Cannot read properties of undefined (reading 'id')", stack: "TypeError: Cannot read properties of undefined (reading 'id')\n    at getUserID (/usr/local/renovate/lib/modules/platfo…"}
      message: "Cannot read properties of undefined (reading 'id')"
      stack: "TypeError: Cannot read properties of undefined (reading 'id')\n    at getUserID (/usr/local/renovate/lib/modules/platform/gitlab/http.ts:10:13)\n    at processTicksAndRejections (node:int…"
    msg: "addAssignees error"
```

**Reply by bbckr** [reply]
I can take a look at this @stdedos

**Reply by bbckr** [reply]
@stdedos Do you have `expandCodeownersGroups` enabled alongside `assigneesFromCodeOwners: true`? Also, are there gitlab groups declared in your CODEOWNERS file? And if so, would it be trying to add that group? 

From the stacktrace, it seems that once we reach `addAssignees`, we could be trying to resolve the assignee Group as a User, hence why there will be no user id available. Another possibility is that we could be trying to resolve the assignee user but having trouble getting the id returned (for whatever the reason may be, need to get more familiar with the Gitlab API to see these scenarios).

This was never the case for when we reach `addReviewers`, since we first try to resolve the reviewer as a user and (if fails) a group. We could try to do the same check for adding Assignees, but depends if Gitlab allows adding Groups to Assignees (which I don't know off the top of my head).


Knowing how your CODEOWNERS file is setup and then the other options you have as enabled will be helpful in this case, and then I can try to dive a little more into later to see if I can get this replicated in tests.

**Reply by bbckr** [reply]
I **_think_** if expandCodeownersGroups was enabled, it may avoid this altogether, but let me know what you have configured

**Reply by stdedos** [reply]
_Sadly_, the project-in-question was discontinued - so there will be no more investigation 😕 

Let's say that the `expandCodeownersGroups` missing might have explained the behavior I was seeing

**Reply by stdedos** [reply]
On a ... small tangent from the ``.

"Somehow", I don't see the setting "being offered as a completion" in IDEA
![image](https://github.com/renovatebot/renovate/assets/133706/b2e786bd-f1d5-4598-a024-09fff6025398)
... but it seems to exist on the schema alright (https://docs.renovatebot.com/renovate-schema.json)

Others are offered though:
![image](https://github.com/renovatebot/renovate/assets/133706/1c126811-f6ac-4238-b23a-add04b161aa2)

Any "off the top of someone's head" ideas? 😕

**Comment by uberflip-jbooth** [comment]
I would like to use `expandCodeOwnersGroups` in Github.  Currently we have some PRs that can auto-merge but get "stuck" and it would be nice to automatically assign one of the users from a CODEOWNERS group to make sure someone goes and gets it unstuck.
